PYI-680: Validate callback url in token endpoint

Wynndow opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->
**Supersedes https://github.com/alphagov/di-ipv-core-back/pull/125**

## Proposed changes

### What changed

* Validates the redirect URL provided by the user in the authorisation endpoint by comparing it to the endpoints registered in the clients configuration.
* Validates the redirect URL provided by the client in the auth grant received by the token endpoint, is the same as that received on the authorisation endpoint.
* Refactors the authorisation request validation to a new class.

### Why did it change

OAuth 2 spec made me do it...

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYI-680](https://govukverify.atlassian.net/browse/PYI-680)

Wynndow committed:
PYI-671: Validate redirect URLs

We've added configuration for each client the core talks to which
includes the valid redirect URLs. This consumes that config and
validates incoming requests on the authorize endpoint.

The changes to the configuration service expect the URLs to be a comma
separated list.

This refactors all the validation logic from the authorize endpoint into
a separate validator class to keep the handler cleaner.

Wynndow committed:
PYI-680: Validate redirect URL in token endpoint

Check that the redirect URL received in the grant by the token
endpoint matches that received in the initial call to the authorisation
endpoint, as per the oauth 2 spec.

This updates the auth endpoint to include the redirect URL in the entry
in the auth codes dynamo DB table. We then compare it with the grant
when we pull that value out in the token endpoint.

Danomanic commented on lambdas/accesstoken/src/test/java/uk/gov/di/ipv/core/accesstoken/AccessTokenHandlerTest.java:
```diff
@@ -166,6 +171,33 @@ void shouldReturn400OWhenInvalidAuthorisationCodeProvided() throws Exception {
     }
 
+    @Test
+    void shouldReturn400WhenRedirectUrlInAuthGrantDoesNotMatchThatSuppliedToAuthEndpoint()
```
Would it be worth adding a quick test to see what happens when `redirect_uri` isn't in the request? Not sure if it would cause a null pointer exception in `redirectUrlsDoNotMatch()`

Wynndow replied:
Nice, good spot. I've pushed up a commit to address it.

Danomanic commented on lib/src/main/java/uk/gov/di/ipv/core/library/validation/AuthRequestValidator.java:
```diff
@@ -0,0 +1,86 @@
+package uk.gov.di.ipv.core.library.validation;
```
I like this a lot. Much better way to validate the request. Maybe something we could use on other Handlers.

Wynndow committed:
PYI-680: Handle null values for redirect url's

It's valid in the spec for the redirect value to be omitted by the
client in the auth request. And if it is, it should also be omitted in
the token request.

At the moment this won't happen as we're validating the presence of the
redirect url in the auth endpoint, but we might handle it differently in
future.
